### ISSUE TITLE
deprecation of setChannel drops support for mTLS

### ISSUE THREAD

**Issue by joske1234** [open-telemetry/opentelemetry-java#4165]
**Is your feature request related to a problem? Please describe.**
The setChannel(ManagedChannel channel) method on OtlpGrpcSpanExporterBuilder is deprecated in the latest release. This method allowed setting up a grpc channel secured by mTLS. With the now suggested setEndpoint() this seems no longer possible.

**Describe the solution you'd like**
Revert the deprecation or include mTLS support for setEndpoint().

**Comment by anuraaga** [comment]
Thanks @joske1234 - note that there is no short or medium term plan to remove the methods so you can continue to use them while needed. We don't want users to need to use gRPC APIs to access useful features or have gaps vs our HTTP exporters so are looking exactly for this feedback. It looks like we need to prioritize https://github.com/open-telemetry/opentelemetry-specification/issues/1363

**Comment by evantorrie** [comment]
Agree with @joske1234 that this was an immediate concern when integrating with the just-released v1.11.0.

We require mutual TLS for communication between any services in our infrastructure, so deprecating the ability to use it (both for JaegergRPC and OTLP gRPC) appears to be going the wrong direction.

Note that the OpenTelemetry Collector project has general support for both receivers/exporters to employ full mutual TLS.

**Comment by anuraaga** [comment]
Thanks @evantorrie for letting us know the importance of mTLS. Strictly speaking, OpenTelemetry doesn't support mTLS as it hasn't been added to the specification - we do have a backdoor via `setChannel` right now but it is not consistent across languages or even exporters within this repo so we'd like to move off from it. I've filed a spec repo to hopefully knock this gap out finally and we would have native support for mTLS after that, ideally in the next release. In the meantime, hopefully it's OK to stick with `@SuppressWarnings("deprecation")` for now.

**Comment by evantorrie** [comment]
@anuraaga Yes, it's perfectly fine to stick with `setChannel` for now. I agree that it should be mentioned in the specification that mTLS should be required.

**Comment by trudyq** [comment]
Hi, 

I don't know if it's the right place or if i should have created a new issue. Please tell me.

I have an otr collector endpoint secured with an oidc auth.
In our Android App we get an oidc jwt token obtained after the user has authenticated with our enterprise login redirection.

So i use `setChannel` to set up a grpc channel in which i inject the user jwt token in custom headers, and especially the refreshed token.

In my case i need the ability :

1. to add interceptors to the channel in OtlpGrpcSpanExporterBuilder or 
2. in OtlpGrpcSpanExporter to be able to modify a header value in the channel.

**Comment by trudyq** [comment]
> @trudyq Can you add the header with the [addHeader(String,String)](https://github.com/open-telemetry/opentelemetry-java/blob/main/exporters/otlp/trace/src/main/java/io/opentelemetry/exporter/otlp/trace/OtlpGrpcSpanExporterBuilder.java#L115-L126) API?

I'm afraid not. With addHeader the <key,value> is set once and is immutable. After my jwt token expired i get a refreshed one and i need to update my header with the new jwt token value.

**Comment by jack-berg** [comment]
I see. I suppose we could address that with something like `addHeader(String key, Supplier<String> valueSupplier)`.

**Comment by trudyq** [comment]
@jack-berg Yes definitely !

**Comment by anuraaga** [comment]
https://github.com/open-telemetry/opentelemetry-specification/issues/1344 is the spec issue for tracking exporter authentication mechanisms. There definitely needs to be some thought put into this topic, until then I would say OpenTelemetry only supports fixed token authentication through that `addHeader` API.

That being said, adding the `Supplier` version even without a spec change doesn't seem that bad. But one thing to keep in mind is that it still only opens up some auth patterns, a "proper" auth interface would probably look something like

```java
interface ExporterCredentials {
  void addCredentials(Map<String, String> requestHeaders, byte[] content, ExporterCredentialsCallback callback);
}

interface ExporterCredentialsCallback {
  void onSuccess(Map<String, String> headers);

  void onFailure(Throwable t);
}
```

I would try throwing together in the spec, but that repo is designed to burn out contributors so hopefully someone else eventually will :)

**Comment by pureklkl** [comment]
Looking at the auth extension at collector, they just let the extension provide the auth interceptors for http/grpc. Maybe it will be an easier way to go?
```
public interface ExporterAuthProvider {

  @Nullable
  Interceptor getOkhttpInterceptor();

  @Nullable
  ClientInterceptor getGrpcInterceptor();
}
```

**Comment by anuraaga** [comment]
The SDKs have to be compatible with whatever libraries are in the user's classpath so we have a stronger requirement to avoid leaking dependencies into public APIs than the collector. Perhaps it means the collector isn't a good reference point after all :) But we wouldn't be exposing okhttp and gRPC points like that for an authorization API, as with everything in OpenTelemetry we need the customization actually spec'd out in a cross-language way.

**Comment by jack-berg** [comment]
Resolved in #4194. Opened #4292 to track the remaining issue of a better solution for header based authentication.
